features/environment.py

```diff
@@ -1,294 +1,340 @@
-                if entry == 'features' or cwd == '/':
-                    break
-            self._patroni_path = cwd
-        return self._patroni_path
+    def _has_started(self):
+        return self._handle and self._handle.pid and self._handle.poll() is None
 
-    def data_dir(self, pg_name):
-        return os.path.join(self.patroni_path, 'data', pg_name)
+    def _is_running(self):
+        return self._has_started()
 
-    def write_label(self, pg_name, content):
-        with open(os.path.join(self.data_dir(pg_name), 'label'), 'w') as f:
-            f.write(content)
+    @abc.abstractmethod
+    def _is_accessible(self):
+        """process is accessible for queries"""
 
-    def read_label(self, pg_name):
-        content = None
-        try:
-            with open(os.path.join(self.data_dir(pg_name), 'label'), 'r') as f:
-                content = f.read()
-        except IOError:
-            return None
-        return content.strip()
+    @abc.abstractmethod
+    def _start(self):
+        """start process"""
+
+    def start(self, max_wait_limit=5):
+        if self._is_running():
+            return True
+
+        self._log = open(os.path.join(self._output_dir, self._name + '.log'), 'a')
+        self._handle = self._start()
+
+        assert self._has_started(), "Process {0} is not running after being started".format(self._name)
 
-    def start(self, pg_name, max_wait_limit=20, tags=None):
-        if not self._is_running(pg_name):
-            if pg_name in self._processes:
-                del self._processes[pg_name]
-            cwd = self.patroni_path
-            self._log[pg_name] = open(os.path.join(self._output_dir, 'patroni_{0}.log'.format(pg_name)), 'a')
-
-            self._config[pg_name] = self._make_patroni_test_config(pg_name, tags=tags)
-
-            p = subprocess.Popen(['coverage', 'run', '--branch', '--source=patroni', '-p', 'patroni.py', self._config[pg_name]],
-                                 stdout=self._log[pg_name], stderr=subprocess.STDOUT, cwd=cwd)
-            if not (p and p.pid and p.poll() is None):
-                assert False, "PostgreSQL {0} is not running after being started".format(pg_name)
-            self._processes[pg_name] = p
-        # wait while patroni is available for queries, but not more than 10 seconds.
         for _ in range(max_wait_limit):
-            if self.query(pg_name, "SELECT 1", fail_ok=True) is not None:
+            if self._is_accessible():
                 break
             time.sleep(1)
         else:
             assert False,\
-                "Patroni instance is not available for queries after {0} seconds".format(max_wait_limit)
+                "{0} instance is not available for queries after {1} seconds".format(self._name, max_wait_limit)
 
-    def stop(self, pg_name, kill=False, timeout=15):
+    def stop(self, kill=False, timeout=15):
+        """ terminate process and wipe out the temp work directory, but only if we actually started it"""
```

Comment:
I think the comment is incorrect, the stop function doesn't try to delete any directories.